### ISSUE TITLE
panic on `(*Stream).getState`

### ISSUE THREAD

**Issue by howardjohn** [grpc/grpc-go#4603]
### What version of gRPC are you using?
1.39 and 1.36
### What version of Go are you using (`go version`)?
1.16
### What operating system (Linux, Windows, …) and version?
Linux
### What did you do?
Connected to a bidi streaming gRPC service.

### What did you expect to see?
Service works correctly

### What did you see instead?
Got a panic instead.

Full details are in https://github.com/istio/istio/issues/33967. I am opening this here because after some analysis on our side, this looks to be more like a gRPC issue than an Istio issue - if that is not the case please let me know.

**Comment by dfawley** [comment]
Unfortunately not much to go on with this one so far.  The only clue I see is that with GRPC logging enabled, the ClientConn was shutting down (or had already shut down) when this occurred.  Otherwise everything looks pretty normal except for the panic.

**Comment by howardjohn** [comment]
@dfawley does this look like it could be an error on our end, or more likely in grpc internals?

Is there any useful information we can collect? I cannot reproduce this but it seems the original reporter is able to

**Comment by dfawley** [comment]
Sorry for the lack of an update here.  We definitely should never panic in our library, so even if you are using it incorrectly, we do have a bug that needs to be fixed.  Is this still happening?  Do you have a way to reproduce it?  Can you provide us with something that can reproduce it?

Can you provide anything else to go on here?  As I mentioned before, it looked like it was possible a shutdown of the `ClientConn` was happening as the panic occurred -- can you confirm this?  Collecting a log of all the operations you're performing when this happens could be helpful in being able to reproduce it.

**Comment by stefan-caraiman** [comment]
Here is another log stack trace where it seems like `grpc` gets a panic :

```
2021-08-12T09:10:12.435014Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --drain-strategy immediate --parent-shutdown-time-s 60 --service-cluster test-pod.test --service-node sidecar~10.1.1.139~test-pod-1628759400-p7hhh.test~test.svc.cluster.local --local-address-ip-version v4 --bootstrap-version 3 --disable-hot-restart --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error --concurrency 2]
2021-08-12T09:10:12.507063Z	info	[core]parsed scheme: ""
2021-08-12T09:10:12.507091Z	info	[core]scheme "" not registered, fallback to default scheme
2021-08-12T09:10:12.507116Z	info	[core]ccResolverWrapper: sending update to cc: {[{istiod.istio-system.svc:15012  <nil> 0 <nil>}] <nil> <nil>}
2021-08-12T09:10:12.507151Z	info	[core]ClientConn switching balancer to "pick_first"
2021-08-12T09:10:12.507165Z	info	[core]Channel switches to new LB policy "pick_first"
2021-08-12T09:10:12.507195Z	info	[core]Subchannel Connectivity change to CONNECTING
2021-08-12T09:10:12.507216Z	info	[core]blockingPicker: the picked transport is not ready, loop back to repick
2021-08-12T09:10:12.507231Z	info	[core]Subchannel picks a new address "istiod.istio-system.svc:15012" to connect
2021-08-12T09:10:12.507273Z	info	[core]pickfirstBalancer: UpdateSubConnState: 0xc00051c800, {CONNECTING <nil>}
2021-08-12T09:10:12.507339Z	info	[core]Channel Connectivity change to CONNECTING
2021-08-12T09:10:12.515030Z	info	[core]Subchannel Connectivity change to READY
2021-08-12T09:10:12.515056Z	info	[core]pickfirstBalancer: UpdateSubConnState: 0xc00051c800, {READY <nil>}
2021-08-12T09:10:12.515061Z	info	[core]Channel Connectivity change to READY
2021-08-12T09:10:12.515147Z	info	[core]Subchannel Connectivity change to CONNECTING
2021-08-12T09:10:12.515183Z	info	xdsproxy	connected to upstream XDS server: istiod.istio-system.svc:15012
2021-08-12T09:10:12.515236Z	info	[core]blockingPicker: the picked transport is not ready, loop back to repick
2021-08-12T09:10:12.515261Z	info	[transport]transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2021-08-12T09:10:12.515273Z	info	[core]pickfirstBalancer: UpdateSubConnState: 0xc00051c800, {CONNECTING <nil>}
2021-08-12T09:10:12.515278Z	info	[core]Channel Connectivity change to CONNECTING
2021-08-12T09:10:12.515288Z	info	[core]Subchannel picks a new address "istiod.istio-system.svc:15012" to connect
2021-08-12T09:10:12.523361Z	info	[core]Subchannel Connectivity change to READY
2021-08-12T09:10:12.523396Z	info	[core]pickfirstBalancer: UpdateSubConnState: 0xc00051c800, {READY <nil>}
2021-08-12T09:10:12.523403Z	info	[core]Channel Connectivity change to READY
2021-08-12T09:10:12.523472Z	info	[core]Subchannel Connectivity change to CONNECTING
2021-08-12T09:10:12.523546Z	info	[transport]transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2021-08-12T09:10:12.523567Z	info	[core]pickfirstBalancer: UpdateSubConnState: 0xc00051c800, {CONNECTING <nil>}
2021-08-12T09:10:12.523575Z	info	[core]Channel Connectivity change to CONNECTING
2021-08-12T09:10:12.523589Z	info	[core]Subchannel picks a new address "istiod.istio-system.svc:15012" to connect
2021-08-12T09:10:12.523627Z	info	[core]Channel Connectivity change to SHUTDOWN
2021-08-12T09:10:12.523644Z	info	[core]Subchannel Connectivity change to SHUTDOWN
2021-08-12T09:10:12.523745Z	info	[core]grpc: addrConn.createTransport failed to connect to {istiod.istio-system.svc:15012 istiod.istio-system.svc <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial tcp: operation was canceled". Reconnecting...
2021-08-12T09:10:12.523998Z	info	[transport]transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2021-08-12T09:10:12.524875Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, transport is closing
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xd8 pc=0xd03952]

goroutine 77 [running]:
google.golang.org/grpc/internal/transport.(*Stream).getState(...)
	google.golang.org/grpc@v1.36.0/internal/transport/transport.go:312
google.golang.org/grpc/internal/transport.(*http2Client).Write(0xc000039340, 0x0, 0xc0005c04e6, 0x5, 0x5, 0xc000786000, 0x3855, 0x3855, 0xc000b37dac, 0x28e97e0, ...)
	google.golang.org/grpc@v1.36.0/internal/transport/http2_client.go:912 +0x1f2
google.golang.org/grpc.(*csAttempt).sendMsg(0xc00040e100, 0x19264e0, 0xc000b80900, 0xc0005c04e6, 0x5, 0x5, 0xc000786000, 0x3855, 0x3855, 0xc000786000, ...)
	google.golang.org/grpc@v1.36.0/stream.go:889 +0x129
google.golang.org/grpc.(*clientStream).SendMsg.func2(0xc00040e100, 0x7f2eaed26f18, 0x50)
	google.golang.org/grpc@v1.36.0/stream.go:754 +0xb3
google.golang.org/grpc.(*clientStream).withRetry(0xc000b39d40, 0xc000204e60, 0xc00062fef0, 0x2914700, 0x0)
	google.golang.org/grpc@v1.36.0/stream.go:634 +0x3df
google.golang.org/grpc.(*clientStream).SendMsg(0xc000b39d40, 0x19264e0, 0xc000b80900, 0x0, 0x0)
	google.golang.org/grpc@v1.36.0/stream.go:760 +0x3a5
github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3.(*aggregatedDiscoveryServiceStreamAggregatedResourcesClient).Send(0xc000974270, 0xc000b80900, 0x1cf3fb8, 0x0)
	github.com/envoyproxy/go-control-plane@v0.9.9-0.20210512163311-63b5d3c536b0/envoy/service/discovery/v3/ads.pb.go:233 +0x49
istio.io/istio/pkg/istio-agent.sendUpstreamWithTimeout.func1(0xc0001244e0)
	istio.io/istio/pkg/istio-agent/xds_proxy.go:687 +0x38
created by istio.io/istio/pkg/istio-agent.sendWithTimeout
	istio.io/istio/pkg/istio-agent/xds_proxy.go:705 +0xdd
```

**Comment by dfawley** [comment]
@stefan-caraiman 

Can you provide any information about what operations are happening on the stream that panics, with timestamps to line the operations up with the ClientConn activity in the logs?  It seems like you are sending on a stream after closing the client.  This would be very unusual, and something you could fix on the application side, even though we obviously shouldn't ever be panicking even in this situation.

**Comment by howardjohn** [comment]
Here is the client code FWIW: https://github.com/istio/istio/blob/6da7a3193970e46ebe6cf748f6411f1de1d286ec/pkg/istio-agent/xds_proxy.go#L387

I think it can hypothetically send after closed due to running send in goroutines, but Idid try to artificially trigger this and didn't see a crash:
```
2021-08-13T16:26:03.557881Z     info    [transport]transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2021-08-13T16:26:16.330672Z     info    [core]parsed scheme: ""
2021-08-13T16:26:16.330709Z     info    [core]scheme "" not registered, fallback to default scheme
2021-08-13T16:26:16.330741Z     info    [core]ccResolverWrapper: sending update to cc: {[{localhost:15012  <nil> 0 <nil>}] <nil> <nil>}
2021-08-13T16:26:16.330758Z     info    [core]ClientConn switching balancer to "pick_first"
2021-08-13T16:26:16.330816Z     info    [core]Channel switches to new LB policy "pick_first"
2021-08-13T16:26:16.330854Z     info    [core]Subchannel Connectivity change to CONNECTING
2021-08-13T16:26:16.330894Z     info    [core]blockingPicker: the picked transport is not ready, loop back to repick
2021-08-13T16:26:16.330904Z     info    [core]pickfirstBalancer: UpdateSubConnState: 0xc0010bc4d0, {CONNECTING <nil>}
2021-08-13T16:26:16.330934Z     info    [core]Channel Connectivity change to CONNECTING
2021-08-13T16:26:16.330929Z     info    [core]Subchannel picks a new address "localhost:15012" to connect
2021-08-13T16:26:16.338359Z     info    [core]Subchannel Connectivity change to READY
2021-08-13T16:26:16.338456Z     info    [core]pickfirstBalancer: UpdateSubConnState: 0xc0010bc4d0, {READY <nil>}
2021-08-13T16:26:16.338507Z     info    [core]Channel Connectivity change to READY
2021-08-13T16:26:16.338716Z     info    xdsproxy        connected to upstream XDS server: localhost:15012
2021-08-13T16:26:16.338950Z     info    [core]Channel Connectivity change to SHUTDOWN
2021-08-13T16:26:16.339043Z     info    [core]Subchannel Connectivity change to SHUTDOWN
2021-08-13T16:26:16.339178Z     info    [transport]transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2021-08-13T16:26:16.339397Z     error   xdsproxy        upstream [19] send error for type url type.googleapis.com/envoy.config.cluster.v3.Cluster: EOF
2021-08-13T09:26:16.339827Z     warning envoy config    StreamAggregatedResources gRPC config stream closed: 1, grpc: the client connection is closing
```

Note: logs from `envoy xyz` are from envoy, `xdsproxy` are logs from our client. The server is basically and XDS proxy from `envoy ---> our server that is crashing ---> upstream XDS server (istiod)`

**Comment by dfawley** [comment]
I don't see any obvious errors in the client code.

I also tried my best to reproduce this from what the logs seem to be indicating.  I added retries and random shutdowns of connections and the ClientConn, with stream accesses happening concurrently.  I am unable to ever get the panic to occur.  It seems pretty clear that there is a bug here (unless it's a hardware issue, which is pretty rare), but I just can't figure out how it happens.

One way to "fix" this would be to add defensive programming everywhere and nil check everything.  However, an invariant has been violated and I'd really prefer to fix the bug that leads to that instead.  That said, if we're unable to reproduce it, we can't fix it.
